kasa/xortransport.py

```diff
@@ -0,0 +1,225 @@
+import asyncio
+import contextlib
+import errno
+import logging
+import socket
+import struct
+from pprint import pformat as pf
+from typing import Dict, Generator, Optional
+
+# When support for cpython older than 3.11 is dropped
+# async_timeout can be replaced with asyncio.timeout
+from async_timeout import timeout as asyncio_timeout
+
+from .deviceconfig import DeviceConfig
+from .exceptions import RetryableException, SmartDeviceException
+from .json import loads as json_loads
+from .protocol import BaseTransport
+
+_LOGGER = logging.getLogger(__name__)
+_NO_RETRY_ERRORS = {errno.EHOSTDOWN, errno.EHOSTUNREACH, errno.ECONNREFUSED}
+_UNSIGNED_INT_NETWORK_ORDER = struct.Struct(">I")
+
+
+class XorTransport(BaseTransport):
+    """Implementation of the Xor encryption transport.
+
+    WIP, currently only to ensure consistent __init__ method signatures
+    for protocol classes.  Will eventually incorporate the logic from
+    TPLinkSmartHomeProtocol to simplify the API and re-use the IotProtocol
+    class.
+    """
+
+    DEFAULT_PORT: int = 9999
+    BLOCK_SIZE = 4
+
+    def __init__(self, *, config: DeviceConfig) -> None:
+        super().__init__(config=config)
+        self.reader: Optional[asyncio.StreamReader] = None
+        self.writer: Optional[asyncio.StreamWriter] = None
+        self.query_lock = asyncio.Lock()
+        self.loop: Optional[asyncio.AbstractEventLoop] = None
+
+    @property
+    def default_port(self):
+        """Default port for the transport."""
+        return self.DEFAULT_PORT
+
+    @property
+    def credentials_hash(self) -> str:
+        """The hashed credentials used by the transport."""
+        return ""
+
+    async def _connect(self, timeout: int) -> None:
+        """Try to connect or reconnect to the device."""
+        if self.writer:
+            return
+        self.reader = self.writer = None
+
+        task = asyncio.open_connection(self._host, self._port)
+        async with asyncio_timeout(timeout):
+            self.reader, self.writer = await task
+            sock: socket.socket = self.writer.get_extra_info("socket")
+            # Ensure our packets get sent without delay as we do all
+            # our writes in a single go and we do not want any buffering
+            # which would needlessly delay the request or risk overloading
+            # the buffer on the device
+            sock.setsockopt(socket.IPPROTO_TCP, socket.TCP_NODELAY, 1)
+
+    async def _execute_send(self, request: str) -> Dict:
+        """Execute a query on the device and wait for the response."""
+        assert self.writer is not None  # noqa: S101
+        assert self.reader is not None  # noqa: S101
+        debug_log = _LOGGER.isEnabledFor(logging.DEBUG)
+        if debug_log:
+            _LOGGER.debug("%s >> %s", self._host, request)
+        self.writer.write(XorEncryption.encrypt(request))
+        await self.writer.drain()
+
+        packed_block_size = await self.reader.readexactly(self.BLOCK_SIZE)
```

Comment:
Is this rather a payload length than block size in this context?

Reply:
Maybe but this is just mirroring what's currently there.  We could rename later

Reply:
Fine by me.